doc/users/explain/api_interfaces.rst

```diff
@@ -0,0 +1,312 @@
+  that will plot the data.
+
+The explicit "Axes" interface
+-----------------------------
+
+The "Axes" interface is how Matplotlib is implemented, and many customizations
+and fine-tuning end up needing to be done at this level.  So even if you
+prefer the higher-level interfaces, it is very useful to understand that this
+interface exists, and how to access it.
+
+In general, using it is as easy as instantiating an instance of a
+`~.matplotlib.figure.Figure` class (``fig`` below), using an
+`~.Figure.add_axes` method (or similar) on that object to create an
+`~.matplotlib.axes.Axes` object (``ax`` below), and then calling drawing methods
+on the Axes (``plot`` in this example):
+
+.. plot::
+   :include-source:
+   :align: center
+
+    import matplotlib.pyplot as plt
+
+    fig = plt.figure()
+    ax = fig.add_axes([0.1, 0.1, 0.8, 0.8])
+    l = ax.plot([1, 2, 3, 4], [0, 0.5, 1, 0.2])
+
+Note that ``plot`` itself also returns an object that can be subsequently
+manipulated.
+
+We call this an "explicit" interface because each object is explicitly
+referenced, and used to make the next object.  Keeping handles to the
+objects is very flexible, and allows us to customize the objects after
+they are created, but before they are displayed.
+
+This is also an example of an "imperative" interface because the user tells
+Matplotlib what to do each step of the way.  Of course Matplotlib has defaults
+for many things (line widths, fonts, colors, etc.), but this interface tells
+Matplotlib how to compose the visualization.  Of course this paradigm is not
+pure, in that ``fig.add_axes()`` encapsulates creating many of the objects on
+an Axes (spines, labels, ticks) and ``ax.plot()`` encapsulates creating a
+line plot.  But it is much more "imperative" than something like
+``data.plot()`` as we will see below.
+
+
+The implicit "pyplot" interface
+-------------------------------
+
+The "pyplot" interface arose from Matlab, where one would just call
+``plot([1, 2, 3, 4], [0, 0.5, 1, 0.2])`` and a figure with axes would be
+created for you.  The `~.matplotlib.pyplot` module shadows most of the
+`~.matplotlib.axes.Axes` plotting methods to give the equivalent of
+the above:
+
+.. plot::
+   :include-source:
+   :align: center
+
+    import matplotlib.pyplot as plt
+
+    plt.plot([1, 2, 3, 4], [0, 0.5, 1, 0.2])
+
+This can be convenient, particularly when doing interactive work or simple
+scripts - it saves at least a line of typing, and saves the user from having
+to think about figure and axes creation.  However, for more complicated scripts
+it can be at more of a disadvantage - imagine you want to loop through a number
+of variables and plot them on separate Axes, it is often more straight-forward
+to create the Axes beforehand.  It can also become ambiguous when you have
+multiple Figures or Axes which one is being acted on.  In general the
+Matplotlib project recommends using the explicit interface, at least for code
+you want to share or preserve.
```

Comment:
I would make this recommendation in a separate section, rather than as an aside here.